unity-renderer/Assets/DCLPlugins/ECS7/ComponentCrdtWriteSystem/ComponentCrdtWriteSystem.cs

```diff
@@ -0,0 +1,141 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using DCL;
+using DCL.Controllers;
+using DCL.CRDT;
+using DCL.Interface;
+using KernelCommunication;
+using BinaryWriter = KernelCommunication.BinaryWriter;
+
+public class ComponentCrdtWriteSystem : IDisposable
+{
+    internal const int BINARY_MSG_MAX_SIZE = 5242880;
+
+    internal readonly Dictionary<string, Queue<CRDTMessage>> queuedMessages = new Dictionary<string, Queue<CRDTMessage>>();
+
+    private readonly IUpdateEventHandler updateEventHandler;
+    private readonly MemoryStream memoryStream;
+    private readonly BinaryWriter binaryWriter;
+    private readonly IWorldState worldState;
+
+    public ComponentCrdtWriteSystem(IUpdateEventHandler updateEventHandler, IWorldState worldState)
+    {
+        this.updateEventHandler = updateEventHandler;
+        this.worldState = worldState;
+        updateEventHandler.AddListener(IUpdateEventHandler.EventType.LateUpdate, ProcessMessages);
+        memoryStream = new MemoryStream();
+        binaryWriter = new BinaryWriter(memoryStream);
+    }
+
+    public void Dispose()
+    {
+        updateEventHandler.RemoveListener(IUpdateEventHandler.EventType.LateUpdate, ProcessMessages);
+        binaryWriter.Dispose();
+        memoryStream.Dispose();
+    }
+
+    public void WriteMessage(string sceneId, long entityId, int componentId, byte[] data)
+    {
+        CRDTMessage message = new CRDTMessage()
+        {
+            key = CRDTUtils.KeyFromIds((int)entityId, componentId),
+            timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
+            data = data
+        };
+
+        // dispatch to scene's crdt executor
+        DispatchToCRDTExecutor(sceneId, message);
+
+        // enqueue messages to send to kernel
+        if (!queuedMessages.TryGetValue(sceneId, out Queue<CRDTMessage> sceneMessages))
+        {
+            sceneMessages = new Queue<CRDTMessage>();
+            queuedMessages.Add(sceneId, sceneMessages);
+        }
+
+        sceneMessages.Enqueue(message);
+    }
+
+    internal void ProcessMessages()
```

Comment:
Should this be an async function to be throttled? maybe we can have a IMessageProcessor so we dont have to decide now

Reply:
Fair enough!